system/datatypes/mailinglist.php

```diff
@@ -1,406 +1,433 @@
+		}
+
+		if ($addAddress) {
+			// add email with status pending
+			/** @var HyphaDomElement $address */
+			$address = $this->xml->createElement(self::FIELD_NAME_ADDRESS);
+			$address->setAttribute(self::FIELD_NAME_EMAIL, $email);
+			$address->setAttribute(self::FIELD_NAME_STATUS, self::ADDRESS_STATUS_PENDING);
+			$address->setAttribute(self::FIELD_NAME_CONFIRM_CODE, $code);
+			$addresses->append($address);
+			$this->xml->saveAndUnlock();
+		} else {
+			$this->xml->unlock();
+		}
+
+		if ($sendEmail) {
+			// send email so that pending subscribed can confirm email address
+			$confirmPath = str_replace('{code}', $code, self::PATH_CONFIRM_CODE);
+			$confirmUrl = $this->constructFullPath($this->pagename . '/' . $confirmPath);
+			$confirmLink = '<a href="' . $confirmUrl . '">' . __('ml-please-confirm-email') . '</a>';
+			$welcomeText = $this->xml->find(self::FIELD_NAME_EMAIL_WELCOME_TEXT)->getHtml();
+			$welcomeText .= '<br><br>' . $confirmLink;
+			$subject = hypha_getTitle() . ' - ' . __('ml-confirmation-email-subject');
+			$this->sendMail($subject, $welcomeText, [$email]);
+
+			notify('success', ucfirst(__('ml-confirmation-mail-sent')));
+		} else {
+			notify('success', ucfirst(__('ml-successfully-subscribed')));
+		}
+
+		// all is success refresh page with success notification
+		return 'reload';
+	}
+
+	private function confirmEmailAction(HyphaRequest $request) {
 		$code = isset($_GET['code']) ? $_GET['code'] : null;
 		if (null == $code) {
 			notify('error', __('missing-arguments'));
-
 			return null;
 		}
 
 		$this->xml->lockAndReload();
 
 		/** @var NodeList $addressCollection */
-		$addressCollection = $this->getDoc()->findXPath('//address[@confirm-code="' . $code . '"]');
+		$xpath = '//' . self::FIELD_NAME_ADDRESS . '[@' . self::FIELD_NAME_CONFIRM_CODE . '="' . $code . '" and @' . self::FIELD_NAME_STATUS . ' != "' . self::ADDRESS_STATUS_UNSUBSCRIBED . '"]';
+		$addressCollection = $this->xml->findXPath($xpath);
 		$address = $addressCollection->first();
 		if (!$address instanceof HyphaDomElement) {
 			$this->xml->unlock();
 			notify('error', __('invalid-code'));
-
 			return null;
 		}
 
-		if ($address->getAttribute('status') == self::ADDRESS_STATUS_CONFIRMED) {
+		if ($address->getAttribute(self::FIELD_NAME_STATUS) == self::ADDRESS_STATUS_CONFIRMED) {
 			$this->xml->unlock();
 			notify('success', ucfirst(__('ml-successfully-subscribed')));
 			return ['redirect', $this->constructFullPath($this->pagename)];
 		}
 
-		$address->setAttribute('status', self::ADDRESS_STATUS_CONFIRMED);
-		$address->setAttribute('unsubscribe-code', $this->constructCode());
+		$address->setAttribute(self::FIELD_NAME_STATUS, self::ADDRESS_STATUS_CONFIRMED);
+		$address->setAttribute(self::FIELD_NAME_UNSUBSCRIBE_CODE, $this->constructCode());
 		/** @var HyphaDomElement $addresses */
-		$addresses = $this->getDoc()->getOrCreate('addresses');
+		$addresses = $this->xml->find(self::FIELD_NAME_ADDRESSES_CONTAINER);
```

Comment:
Should this use find rather than getOrCreate? With find, it will look recursively, which is not what is intended I think? Maybe this should use `get()` instead?

Reply:
Updated this troughtout the class